test/common/network/connection_impl_test.cc

```diff
@@ -870,6 +887,173 @@ TEST_P(ConnectionImplTest, EmptyReadOnCloseTest) {
+  EXPECT_CALL(*transport_socket, canFlushClose()).WillOnce(Return(true));
+
+  std::unique_ptr<Network::ConnectionImpl> server_connection(new Network::ConnectionImpl(
+      dispatcher, std::make_unique<ConnectionSocketImpl>(0, nullptr, nullptr),
+      TransportSocketPtr(transport_socket), true));
+  server_connection->setDelayedCloseTimeout(std::chrono::milliseconds(500));
+
+  NiceMockConnectionStats stats;
+  server_connection->setConnectionStats(stats.toBufferStats());
+  EXPECT_CALL(stats.delayed_close_timeouts_, inc()).Times(1);
+
+  Buffer::OwnedImpl data("data");
+  server_connection->write(data, false);
+  // Data is pending in the write buffer, which will trigger the FlushWrite close to go into delayed
+  // close processing.
+  server_connection->close(ConnectionCloseType::FlushWrite);
+
+  // Issue the delayed close callback to ensure connection is closed.
+  timer->callback_();
+}
+
+// Test that a FlushWriteAndDelay close causes Envoy to flush the write and wait for the client/peer
+// to close (until a configured timeout which is not expected to trigger in this test).
+TEST_P(ConnectionImplTest, FlushWriteAndDelayCloseTest) {
+  setUpBasicConnection();
+  connect();
+  // Set a very high timeout value to prevent flaking. We are testing the common case where the
+  // timeout does not trigger.
+  server_connection_->setDelayedCloseTimeout(std::chrono::milliseconds(50000));
+
+  std::shared_ptr<MockReadFilter> client_read_filter(new NiceMock<MockReadFilter>());
+  client_connection_->addReadFilter(client_read_filter);
+
+  NiceMockConnectionStats stats;
+  server_connection_->setConnectionStats(stats.toBufferStats());
+
+  Buffer::OwnedImpl data("Connection: Close");
+  server_connection_->write(data, false);
+
+  EXPECT_CALL(*client_read_filter, onData(BufferStringEqual("Connection: Close"), false))
+      .Times(1)
+      .WillOnce(InvokeWithoutArgs([&]() -> FilterStatus {
+        client_connection_->close(ConnectionCloseType::NoFlush);
+        return FilterStatus::StopIteration;
+      }));
+
+  // Client closes the connection so delayed close timer on the server conn should not fire.
+  EXPECT_CALL(stats.delayed_close_timeouts_, inc()).Times(0);
+  EXPECT_CALL(client_callbacks_, onEvent(ConnectionEvent::LocalClose)).Times(1);
+  EXPECT_CALL(server_callbacks_, onEvent(ConnectionEvent::RemoteClose))
+      .Times(1)
+      .WillOnce(Invoke([&](Network::ConnectionEvent) -> void { dispatcher_->exit(); }));
+  server_connection_->close(ConnectionCloseType::FlushWriteAndDelay);
+  dispatcher_->run(Event::Dispatcher::RunType::Block);
+}
+
+// Test that a FlushWriteAndDelay close triggers a timeout which forces Envoy to close the
+// connection when a client has not issued a close within the configured interval.
+TEST_P(ConnectionImplTest, FlushWriteAndDelayCloseTimerTriggerTest) {
+  setUpBasicConnection();
+  connect();
+  server_connection_->setDelayedCloseTimeout(std::chrono::milliseconds(500));
```

Comment:
Same question about flakes here. How do we know 500ms is the right number?

Reply:
This is safe. Since the client connection does not issue a close() during this test, the timer always triggers. I've added a clarifying comment and reduced the timeout value since there is no need to wait so long.